backend/testsuite/io_integration_test.go

```diff
@@ -0,0 +1,557 @@
+			s.testFileOperations(s.localDir)
+		})
+	}
+	for scheme, location := range s.testLocations {
+		s.T().Run(scheme, func(t *testing.T) {
+			s.testFileOperations(location.URI())
+		})
+	}
+}
+
+type TestCase struct {
+	description       string
+	sequence          string
+	fileAlreadyExists bool
+	expectFailure     bool
+	expectedResults   string
+}
+
+// unless seek or read is called first, writes should replace a file (not edit)
+
+func (s *ioTestSuite) testFileOperations(testPath string) {
+	testCases := []TestCase{
+		// Read, Close file
+		{
+			"Read, Close, file exists",
+			"R(all);C()",
+			true,
+			false,
+			"some text",
+		},
+		{
+			"Read, CLose, file does not exist",
+			"R(all);C()",
+			false,
+			true,
+			"",
+		},
+
+		// Read, Seek, Read, Close
+		{"Read, Seek, Read, Close, file exists",
+			"R(4);S(0,0);R(4);C()",
+			true,
+			false,
+			"some text",
+		},
+
+		// Write, Close
+		{
+			"Write, Close, file does not exist",
+			"W(abc);C()",
+			false,
+			false,
+			"abc",
+		},
+		{
+			"Write, Close, file exists",
+			"W(abc);C()",
+			true,
+			false,
+			"abc",
+		},
+
+		// Write, Seek, Write, Close
+		{
+			"Write, Seek, Write, Close, file does not exist",
+			"W(this and that);S(0,0);W(that);C()",
+			false,
+			false,
+			"that and that",
+		},
+		{
+			"Write, Seek, Write, Close, file exists",
+			"W(this and that);S(0,0);W(that);C()",
+			true,
+			false,
+			"that and that",
+		},
+
+		// Seek
+		{
+			"Seek, Close - file does not exist",
+			"S(2,0);C()",
+			false,
+			true,
+			"",
+		},
+		{
+			"Seek, Close - file exists",
+			"S(2,0);C()",
+			true,
+			false,
+			"some text",
+		},
+		{
+			"Seek, Write, Close, file exists",
+			"S(5,0);W(new text);C()",
+			true,
+			false,
+			"some new text",
+		},
+
+		// Seek, Read, Close
+		{
+			"Seek, Read, Close, file does not exist",
+			"S(5,0);R(4);C()",
+			false,
+			true,
+			"",
+		},
+		{
+			"Seek, Read, Close, file exists", // this ends up being effectively a no-op to the test results but does in-fact work
+			"S(5,0);R(4);C()",
+			true,
+			false,
+			"some text",
+		},
+
+		// Read, Write, Close
+		{
+			"Read, Write, Close, file does not exist",
+			"R(5);W(new text);C()",
+			false,
+			true,
+			"",
+		},
+		{
+			"Read, Write, Close, file exists",
+			"R(5);W(new text);C()",
+			true,
+			false,
+			"some new text",
+		},
+
+		// Read, Seek, Write, Close
+		{
+			"Read, Seek, Write, Close, file does not exist",
+			"R(2);S(3,1);W(new text);C()",
+			false,
+			true,
+			"",
+		},
+		{
+			"Read, Seek, Write, Close, file exists",
+			"R(2);S(3,1);W(new text);C()",
+			true,
+			false,
+			"some new text",
+		},
+
+		// Write, Seek, Read, Close
+		{
+			"Write, Seek, Read, Close, file does not exist",
+			"W(new text);S(0,0);R(5);C()",
+			false,
+			false,
+			"new text",
+		},
+		{
+			"Write, Seek, Read, Close, file exists",
+			"W(new text);S(0,0);R(5);C()",
+			true,
+			false,
+			"new text",
+		},
+	}
```

Comment:
These tests cases use a mini-DSL to perform read/write/seek/close functions in various sequences.  It was surprising how little the SFTP implementation had a to change to match the tests.  Basically, only changes to a couple of flags and (making Read() and Write() use RW (vs read-only or write-only), and the conditional Truncate.

Reply:
Nice! I like the repeatable nature of this.

Reply:
I might be missing some context around how these get executed. Is it expected that contributors add test cases here, or do we expect contributors to run this test suite specifically? Our README doc doesn't really call out how to perform integration tests.

Reply:
Yeah, I actually built these tests as part of a different PR for S3 streaming (vs saving to memory) but needed to "know" how to handle various sequences first.  My intent is to add each backend to the CI (with gh secrets).  Locally, you can run it the same way you do with the earlier (spaghetti-like) backend integration tests.
See https://github.com/C2FO/vfs/blob/master/backend/testsuite/doc.go